File: backend/app/schemas/pipeline.py
from datetime import datetime

from pydantic import BaseModel

from app.models.pipeline_run import PipelineStage, PipelineStatus
from app.models.validation import ValidationCategory


class PipelineRunResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    article_id: int
    current_stage: PipelineStage
    status: PipelineStatus
    started_at: datetime | None
    completed_at: datetime | None
    error_message: str


class PipelineStartRequest(BaseModel):
    article_id: int
    auto_gate_one: bool = False
    format_id: str | None = None


class ValidationResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    category: ValidationCategory
    item: str
    passed: bool
    score: float
    message: str


class ValidationSummary(BaseModel):
    total: int
    passed: int
    failed: int
    score: float
    by_category: dict[str, dict[str, int]]

File: frontend/src/types/pipeline.ts
export type PipelineMode =
  | "idle"
  | "research"
  | "outline"
  | "generate"
  | "validate"
  | "gate2"
  | "published";

export type EventType =
  | "stage_start"
  | "stage_complete"
  | "stage_progress"
  | "stage_error"
  | "gate_pending"
  | "pipeline_complete"
  | "pipeline_error";

export interface PipelineEvent {
  event_type: EventType;
  stage: string;
  message: string;
  data: Record<string, unknown>;
}

export interface ValidationItem {
  category: "style" | "seo" | "aeo" | "geo";
  item: string;
  passed: boolean;
  score: number;
  message: string;
}

export interface ValidationSummary {
  total: number;
  passed: number;
  failed: number;
  score: number;
  by_category: Record<string, { total: number; passed: number }>;
}

export interface PipelineRun {
  id: number;
  article_id: number;
  current_stage: string;
  status: string;
  error_message?: string;
  started_at: string;
  completed_at: string | null;
}
